nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java

```diff
@@ -0,0 +1,451 @@
+
+        if (kuduSession.getFlushMode() == SessionConfiguration.FlushMode.AUTO_FLUSH_BACKGROUND) {
+            rowErrors.addAll(Arrays.asList(kuduSession.getPendingErrors().getRowErrors()));
+        } else {
+            responses.stream()
+                    .filter(OperationResponse::hasRowError)
+                    .map(OperationResponse::getRowError)
+                    .forEach(rowErrors::add);
+        }
+    }
+
+    protected KerberosUser loginKerberosUser(final String principal, final String keytab) throws LoginException {
+        final KerberosUser kerberosUser = new KerberosKeytabUser(principal, keytab);
+        kerberosUser.login();
+        return kerberosUser;
+    }
+
+    protected abstract void doOnTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException;
+
+    @OnStopped
+    public void shutdown() throws Exception {
+        try {
+            if (this.asyncKuduClient != null) {
+                getLogger().debug("Closing KuduClient");
+                this.asyncKuduClient.close();
+                this.asyncKuduClient  = null;
+            }
+        } finally {
+            if (kerberosUser != null) {
+                kerberosUser.logout();
+                kerberosUser = null;
+            }
+        }
+    }
+
+    protected void scan(ProcessContext context, ProcessSession session, KuduTable kuduTable, String predicates, List<String> projectedColumnNames, ResultHandler handler) throws Exception {
+        AsyncKuduScanner.AsyncKuduScannerBuilder scannerBuilder = this.asyncKuduClient.newScannerBuilder(kuduTable);
+        final String[] arrayPredicates = (predicates == null || predicates.isEmpty() ? new String[0] : predicates.split(","));
+
+        for(String column : arrayPredicates){
+            if (column.contains("=")) {
+                final String[] parts = column.split("=");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1], "EQUAL");
+            } else if(column.contains(">")) {
+                final String[] parts = column.split(">");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1], "GREATER");
+            } else if(column.contains("<")) {
+                final String[] parts = column.split("<");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1], "LESS");
+            } else if(column.contains(">=")) {
+                final String[] parts = column.split(">=");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1], "GREATER_EQUAL");
+            } else if(column.contains("<=")) {
+                final String[] parts = column.split("<=");
+                addPredicate(scannerBuilder, kuduTable, parts[0], parts[1], "LESS_EQUAL");
+            }
+        }
+
+        if(!projectedColumnNames.isEmpty()){
+            scannerBuilder.setProjectedColumnNames(projectedColumnNames);
+        }
+
+        AsyncKuduScanner scanner = scannerBuilder.build();
```

Comment:
Can you upgrade to the Kudu 1.10.0 client and use a Java for each loop on the scanner? This will simplify scanning and handle scanner keep alive calls for you.

Reply:
I did not understand this

Reply:
In Kudu 1.10 instead of doing `while (scanner.hasMoreRows())`, you can do:

```
for (RowResult row : scanner) {
...
}
```

Though I think you need to use the standard KuduScanner, which I would recommend anyway. See above where I suggest building a standard KuduClient instead of the async one.